compiler/rustc_mir_transform/src/gvn.rs

```diff
@@ -1325,6 +1387,10 @@ impl<'tcx> VnState<'_, 'tcx> {
         assert!(!value.may_have_provenance(self.tcx, op.layout.size));
 
         let const_ = Const::Val(value, op.layout.ty);
+        // Cache the propagated constant.
+        if let Some(new_index) = self.insert_constant(const_) {
+            self.values.swap_indices(index.as_usize(), new_index.as_usize());
+        }
```

Comment:
No. This breaks many invariants of the pass. `new_index` may not be newly created. The caching is what `evaluated` does, or should use a new array if required.

Reply:
I was wrong. This broke the invariants of the pass. Reverted.